## Accessing your data from the database

- Please follow the steps in this notebook to have access to the dataset. 
- If you encounter any challenges please leave an issue on this repo here on GitHub

### Steps to take to use environment variables as opposed to credentials literals

1. Install pyodbc  - a package for creating connection strings to your remote database server
2. Install python-dotenv - a package for creating environment variables that will help you hide sensitve configuration informantion such as database credentials and API keys
3. Import all the necessary libraies
   1. pyodbc (for creating a connection)
   2. python-dotenv (loading environment variables)
   3. os (for accessing the environement variables using the load_env function. This is not needed if you use the dotenv_values function instead)
4. Now create a file called .env in the root of your project folder (Note, the file name begins with a dot)
5. In the .env file, put all your sensitive information like server name, database name, username, and password

Example

   - SERVER='server_name_here'
   - DATABASE='database_name_here'
   - USERNAME='username_here'
   - PASSWORD='password_here'


6. Next create a .gitignore file (a new file with the name `.gitignore`. Note that gitignore file names begin with a dot)
7. Open the .gitignore file and type in the name of the .env file we just created like this "/.env". This will prevent git from tracking that file. Essesntially any file name in the gitignore file will be ignored by git and won't be checked into the repository
8. Create a connection by accessing your connection string with your defined environment variables

## Understanding the Business

-Venturing into the Indian start-ups ecosystem
-To investigate the ecosystem and propose the best course of action

-We will analyze funding recieved by start-ups in India from 2018 to 2021.

-We will seek to ask the following questions to help us propose the best cousrse of action.

##### 1) what is the average number of start-ups registered in Indian each year

##### 2) what is the average amount of funding recieved by a start-up per year

##### 3) Which industry do these start-ups belong

##### 4) How much funding in total is alocated to the industries

##### 5) Which start-ups recieved most funding and the industries they belong

##### 6) which start-ups survived after their first year of operation and the industry they belong

##### 7) which start-ups survived after the second year of operation and the industry they belong

##### 8) Which industries have the most successful start-ups

##### 9) What the loccations of the industrie that recieve most funding

##### 10) What is the Location of Industries that  



#### Step 1 and 2 - Install pyodbc and python-dotenv

In [24]:
#!pip install pyodbc  
#%pip install python-dotenv 

#!pip install pymssql
#!pip install pypyodbc

In [23]:
#pip install --upgrade pyodbc

#### Step 3 - Import all the necessary packages

In [1]:
import pyodbc
import pymssql
import pypyodbc

from dotenv import dotenv_values    #import the dotenv_values function from the dotenv package
import pandas as pd
import numpy as np


import warnings 

warnings.filterwarnings('ignore')


ModuleNotFoundError: No module named 'pyodbc'

#### Step 4 - Create your .env file in the root of your project

#### Step 5 - In the .env file, put all your sensitive information like server name, password etc


#### Step 6 & 7 - Next create a .gitignore file and type '/.env' file we just created. This will prevent git from tracking that file.

#### Step 8 - Create a connection by accessing your connection string with your defined environment variables

In [2]:
# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values('`.env`')

# Get the values for the credentials you set in the '.env' file
server = environment_variables.get("SERVER")
database = environment_variables.get("DATABASE")
username = environment_variables.get("USERNAME")
password = environment_variables.get("PASSWORD")
driver = environment_variables.get("DRIVER")


In [3]:
# Create a connection string
connection_string = f'DRIVER={driver};SERVER={server};DATABASE={database};UID={username};PWD={password}'


In [4]:
# Use the connect method of the pyodbc library and pass in the connection string.
# This will connect to the server and might take a few seconds to be complete. 
# Check your internet connection if it takes more time than necessary

connection = pyodbc.connect(connection_string)


In [7]:
# Now the sql query to get the data is what what you see below. 
# Note that you will not have permissions to insert delete or update this database table. 

query = "SELECT * FROM dbo.LP1_startup_funding2020"

df_fund2020 = pd.read_sql(query, connection)
df_fund2020.head(3)

Company_Brand  Founded HeadQuarter              Sector  \
0    Aqgromalin   2019.0     Chennai            AgriTech   
1      Krayonnz   2019.0   Bangalore              EdTech   
2  PadCare Labs   2018.0        Pune  Hygiene management   

                                        What_it_does  \
0                       Cultivating Ideas for Profit   
1  An academy-guardian-scholar centric ecosystem ...   
2   Converting bio-hazardous waste to harmless waste   

                           Founders         Investor    Amount     Stage  \
0   Prasanna Manogaran, Bharani C L  Angel investors  200000.0      None   
1  Saurabh Dixit, Gurudutt Upadhyay  GSF Accelerator  100000.0  Pre-seed   
2                   Ajinkya Dhariya   Venture Center       NaN  Pre-seed   

  column10  
0     None  
1     None  
2     None

In [16]:
query1 = 'Select * from dbo.LP1_startup_funding2021'

df_fund2021 = pd.read_sql(query1, connection)
df_fund2021.head(3)

Company_Brand  Founded HeadQuarter      Sector  \
0  Unbox Robotics   2019.0   Bangalore  AI startup   
1          upGrad   2015.0      Mumbai      EdTech   
2     Lead School   2012.0      Mumbai      EdTech   

                                        What_it_does  \
0  Unbox Robotics builds on-demand AI-driven ware...   
1     UpGrad is an online higher education platform.   
2  LEAD School offers technology based school tra...   

                                            Founders  \
0                        Pramod Ghadge, Shahid Memon   
1  Mayank Kumar, Phalgun Kompalli, Ravijot Chugh,...   
2                         Smita Deorah, Sumeet Mehta   

                                   Investor        Amount         Stage  
0               BEENEXT, Entrepreneur First    $1,200,000  Pre-series A  
1  Unilazer Ventures, IIFL Asset Management  $120,000,000          None  
2          GSV Ventures, Westbridge Capital   $30,000,000      Series D

Next, get data from other sources and concatenate (Depends on the project) to perform your analysis

ALL THE BEST!!!

In [66]:

# Print the first few rows of the DataFrame
#print(data2.head())
#converting the sql extracted data to csv respectively
data.to_csv("startup_funding2020.csv", index=False)
data1.to_csv("startup_funding2021.csv", index=False)

In [9]:
df_fund2018=pd.read_csv('startup_funding2018.csv')

In [10]:
df_fund2019=pd.read_csv('startup_funding2019.csv')


In [11]:
df_fund2018.head(3)

Company Name                                           Industry  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   

  Round/Series       Amount                     Location  \
0         Seed       250000  Bangalore, Karnataka, India   
1         Seed  ₹40,000,000   Mumbai, Maharashtra, India   
2     Series A  ₹65,000,000      Gurgaon, Haryana, India   

                                       About Company  
0  TheCollegeFever is a hub for fun, fiesta and f...  
1  A startup which aggregates milk from dairy far...  
2          Leading Online Loans Marketplace in India

In [13]:
df_fund2019.head(3)

Company/Brand  Founded HeadQuarter     Sector  \
0  Bombay Shaving      NaN         NaN  Ecommerce   
1       Ruangguru   2014.0      Mumbai     Edtech   
2        Eduisfun      NaN      Mumbai     Edtech   

                                        What it does  \
0         Provides a range of male grooming products   
1  A learning platform that provides topic-based ...   
2            It aims to make learning fun via games.   

                                Founders  \
0                     Shantanu Deshpande   
1  Adamas Belva Syah Devara, Iman Usman.   
2                          Jatin Solanki   

                                         Investor     Amount($)          Stage  
0                            Sixth Sense Ventures    $6,300,000            NaN  
1                                General Atlantic  $150,000,000       Series C  
2  Deepak Parekh, Amitabh Bachchan, Piyush Pandey   $28,000,000  Fresh funding

In [14]:
df_fund2020.head(3)

Company_Brand  Founded HeadQuarter      Sector  \
0  Unbox Robotics   2019.0   Bangalore  AI startup   
1          upGrad   2015.0      Mumbai      EdTech   
2     Lead School   2012.0      Mumbai      EdTech   

                                        What_it_does  \
0  Unbox Robotics builds on-demand AI-driven ware...   
1     UpGrad is an online higher education platform.   
2  LEAD School offers technology based school tra...   

                                            Founders  \
0                        Pramod Ghadge, Shahid Memon   
1  Mayank Kumar, Phalgun Kompalli, Ravijot Chugh,...   
2                         Smita Deorah, Sumeet Mehta   

                                   Investor        Amount         Stage  
0               BEENEXT, Entrepreneur First    $1,200,000  Pre-series A  
1  Unilazer Ventures, IIFL Asset Management  $120,000,000          None  
2          GSV Ventures, Westbridge Capital   $30,000,000      Series D

In [17]:
df_fund2021.head(3)

Company_Brand  Founded HeadQuarter      Sector  \
0  Unbox Robotics   2019.0   Bangalore  AI startup   
1          upGrad   2015.0      Mumbai      EdTech   
2     Lead School   2012.0      Mumbai      EdTech   

                                        What_it_does  \
0  Unbox Robotics builds on-demand AI-driven ware...   
1     UpGrad is an online higher education platform.   
2  LEAD School offers technology based school tra...   

                                            Founders  \
0                        Pramod Ghadge, Shahid Memon   
1  Mayank Kumar, Phalgun Kompalli, Ravijot Chugh,...   
2                         Smita Deorah, Sumeet Mehta   

                                   Investor        Amount         Stage  
0               BEENEXT, Entrepreneur First    $1,200,000  Pre-series A  
1  Unilazer Ventures, IIFL Asset Management  $120,000,000          None  
2          GSV Ventures, Westbridge Capital   $30,000,000      Series D

#### We will create year column for each dataframe before concatenating to helps us identify which year funding was awarded

In [55]:
df_fund2018['Year']=2018
df_fund2019['Year']=2019
df_fund2020['Year']=2020
df_fund2021['Year']=2021

In [56]:
df_fund2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Company_Brand  526 non-null    object
 1   Sector         526 non-null    object
 2   Stage          526 non-null    object
 3   Amount         526 non-null    object
 4   Location       526 non-null    object
 5   About Company  526 non-null    object
 6   Year           526 non-null    int64 
dtypes: int64(1), object(6)
memory usage: 28.9+ KB


In [57]:
df_fund2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89 entries, 0 to 88
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  89 non-null     object 
 1   Founded        60 non-null     float64
 2   Location       70 non-null     object 
 3   Sector         84 non-null     object 
 4   About Company  89 non-null     object 
 5   Founders       86 non-null     object 
 6   Investor       89 non-null     object 
 7   Amount         89 non-null     object 
 8   Stage          43 non-null     object 
 9   Year           89 non-null     int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 7.1+ KB


In [58]:
df_fund2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1209 entries, 0 to 1208
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1209 non-null   object 
 1   Founded        1208 non-null   float64
 2   Location       1208 non-null   object 
 3   Sector         1209 non-null   object 
 4   About Company  1209 non-null   object 
 5   Founders       1205 non-null   object 
 6   Investor       1147 non-null   object 
 7   Amount         1206 non-null   object 
 8   Stage          781 non-null    object 
 9   Year           1209 non-null   int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 94.6+ KB


In [59]:
df_fund2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1209 entries, 0 to 1208
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1209 non-null   object 
 1   Founded        1208 non-null   float64
 2   Location       1208 non-null   object 
 3   Sector         1209 non-null   object 
 4   About Company  1209 non-null   object 
 5   Founders       1205 non-null   object 
 6   Investor       1147 non-null   object 
 7   Amount         1206 non-null   object 
 8   Stage          781 non-null    object 
 9   Year           1209 non-null   int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 94.6+ KB


### Observations

- 'Company Name' in df_fund2018 is the same as 'Company/Brand' in df_fund2019, same as 'Company_Brand' in df_fund2020 and same as 'Company_Brand' in df_fund2021

       - we will change Company Name in df_fund2018 to Company_Brand and Company/Brand in df_fund2019 will be change to Company_Brand.

- Founded is in all dataframes except df_fund2018

- We also observed that df_fund2018 contains Industry where as the remaining dataframes contains Sector, however after checking, we noticed the contain similar values.

       - we change Industry in df_fund2018 to Sector
    
- We observed that Location and headquater looks the same, we have Headquarter in all dataframes except df_fund2018, which contains Location
    
    - We will change Headquarter to Location
    
- We observed stage in df_fund2019, df_fund2020, df_fund2021 have the same values as Round/Series in df_fund2018
    
    - we will change Round/Series in df_fund2018 to stage
    
- we observed Amount in df_fund2019 is Amount($), where as all other dataframes contain Amount
    - we will change Amount($) to Amount
    
- we observed About Company in df_fund2018 has similar values as What_it_does in the remaining dataframes, in df_fund2019 there are no underscores
    - we change all to About Company
    

- we observe Founders is in all dataframes except df_fund2018

- we observe Investors is in all dataframes except df_fund2018

- We observed df_fund2020 contains Column10 with null values

    - we will drop column10




In [60]:
#Renaming the Columns for easy concatenation

df_fund2018=df_fund2018.rename(columns={'Company Name': 'Company_Brand'})
df_fund2018=df_fund2018.rename(columns={'Industry': 'Sector'})
df_fund2018=df_fund2018.rename(columns={'Round/Series': 'Stage'})
df_fund2019=df_fund2019.rename(columns={'Company/Brand': 'Company_Brand'})
df_fund2019=df_fund2019.rename(columns={'HeadQuarter': 'Location'})
df_fund2019=df_fund2019.rename(columns={'Amount($)': 'Amount'})
df_fund2019=df_fund2019.rename(columns={'What it does': 'About Company'})
df_fund2020=df_fund2020.rename(columns={'HeadQuarter': 'Location'})
df_fund2020=df_fund2020.rename(columns={'What_it_does': 'About Company'})
df_fund2021=df_fund2021.rename(columns={'HeadQuarter': 'Location'})
df_fund2021=df_fund2021.rename(columns={'What_it_does': 'About Company'})

In [64]:
#droping column10 in df_fund2020
df_fund2020=df_fund2020.drop('column10', axis=1)


In [65]:
#Concatenating Dataframes
df = pd.concat([df_fund2018, df_fund2019, df_fund2020, df_fund2021], ignore_index=True)

In [66]:
df.head()

Company_Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage       Amount                          Location  \
0      Seed       250000       Bangalore, Karnataka, India   
1      Seed  ₹40,000,000        Mumbai, Maharashtra, India   
2  Series A  ₹65,000,000           Gurgaon, Haryana, India   
3     Angel      2000000       Noida, Uttar Pradesh, India   
4      Seed            —  Hyderabad, Andhra Pradesh, India   

                                       About Company  Year  Founded Founders  \
0  TheCollegeFever is a hub for fun, fiesta and f...  2018      NaN      NaN   
1  A startup which aggregates milk from dairy far...  2018      NaN      NaN   
2          Leading Online Loans Marketplace in India  2018      NaN      NaN   
3  PayMe India is an innovative FinTech organizat...  2018      NaN      NaN   
4  Eunimart is a one stop solution for merchants ...  2018      NaN      NaN   

  Investor  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

In [67]:
df.shape

(3033, 10)

#### Observations

- After combining the datasets from the various sources, we got 3033 rows and 9 columns

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3033 entries, 0 to 3032
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  3033 non-null   object 
 1   Sector         3028 non-null   object 
 2   Stage          2131 non-null   object 
 3   Amount         3027 non-null   object 
 4   Location       3012 non-null   object 
 5   About Company  3033 non-null   object 
 6   Year           3033 non-null   int64  
 7   Founded        2476 non-null   float64
 8   Founders       2496 non-null   object 
 9   Investor       2383 non-null   object 
dtypes: float64(1), int64(1), object(8)
memory usage: 237.1+ KB


#### Observations

- We observed that the Amount column is indicating as object type insterd of float, we will investigate further and change to float
- all other object types are correctly specified.
- Year and Founded are suposed to be datetime type but are indicating int and float types, we will change them to datetime

In [72]:
#converting Year and Founded datatypes to datetime

df['Year']=pd.to_datetime(df['Year'], format = '%Y')
df['Founded']=pd.to_datetime(df['Founded'], format = '%Y')

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3033 entries, 0 to 3032
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Company_Brand  3033 non-null   object        
 1   Sector         3028 non-null   object        
 2   Stage          2131 non-null   object        
 3   Amount         3027 non-null   object        
 4   Location       3012 non-null   object        
 5   About Company  3033 non-null   object        
 6   Year           3033 non-null   datetime64[ns]
 7   Founded        2476 non-null   datetime64[ns]
 8   Founders       2496 non-null   object        
 9   Investor       2383 non-null   object        
dtypes: datetime64[ns](2), object(8)
memory usage: 237.1+ KB


In [73]:
df['Amount'].unique()

array(['250000', '₹40,000,000', '₹65,000,000', '2000000', '—', '1600000',
       '₹16,000,000', '₹50,000,000', '₹100,000,000', '150000', '1100000',
       '₹500,000', '6000000', '650000', '₹35,000,000', '₹64,000,000',
       '₹20,000,000', '1000000', '5000000', '4000000', '₹30,000,000',
       '2800000', '1700000', '1300000', '₹5,000,000', '₹12,500,000',
       '₹15,000,000', '500000', '₹104,000,000', '₹45,000,000', '13400000',
       '₹25,000,000', '₹26,400,000', '₹8,000,000', '₹60,000', '9000000',
       '100000', '20000', '120000', '₹34,000,000', '₹342,000,000',
       '$143,145', '₹600,000,000', '$742,000,000', '₹1,000,000,000',
       '₹2,000,000,000', '$3,980,000', '$10,000', '₹100,000',
       '₹250,000,000', '$1,000,000,000', '$7,000,000', '$35,000,000',
       '₹550,000,000', '$28,500,000', '$2,000,000', '₹240,000,000',
       '₹120,000,000', '$2,400,000', '$30,000,000', '₹2,500,000,000',
       '$23,000,000', '$150,000', '$11,000,000', '₹44,000,000',
       '$3,240,000', '₹60